src/domain/synthetics/trade/useTradeboxState.ts

```diff
@@ -621,62 +648,131 @@ function setToTokenAddressUpdaterBuilder(
  * It must set any suitable token and market addresses combination in case it is not correct.
  */
 function fallbackPositionTokens(
-  newState: StoredTradeOptions,
-  availableSwapTokenAddresses: string[],
-  marketAddressIndexTokenMap: { [marketAddress: string]: string }
-): StoredTradeOptions {
-  const needFromUpdate = !availableSwapTokenAddresses.find((t) => t === newState.tokens.fromTokenAddress);
-  const nextFromTokenAddress =
-    needFromUpdate && availableSwapTokenAddresses.length
-      ? availableSwapTokenAddresses[0]
-      : newState.tokens.fromTokenAddress;
-
-  if (nextFromTokenAddress && nextFromTokenAddress !== newState.tokens.fromTokenAddress) {
-    newState = {
-      ...newState,
-      tokens: {
-        ...newState.tokens,
-        fromTokenAddress: nextFromTokenAddress,
-      },
-    };
+  chainId: number,
+  prevState: StoredTradeOptions,
+  nextState: StoredTradeOptions,
+  allowedPayTokens: string[],
+  allowedMarkets: MarketInfo[]
+) {
+  const longOrShort = nextState.tradeType === TradeType.Long ? "long" : "short";
+
+  const allowedPayTokensSet = new Set(allowedPayTokens);
+  const allowedIndexTokens = new Set(allowedMarkets.map((m) => m.indexToken.address));
+
+  const marketsMap = keyBy(allowedMarkets, (m) => m.marketTokenAddress);
+
+  const nextPayTokenAddress = nextState.tokens.fromTokenAddress;
+  const nextIndexTokenAddress = nextState.tokens.indexTokenAddress;
+  const nextMarketTokenAdress = nextIndexTokenAddress && nextState.markets[nextIndexTokenAddress]?.[longOrShort];
+
+  const isNextPayTokenValid = nextPayTokenAddress && allowedPayTokensSet.has(nextPayTokenAddress);
+  const isNextIndexTokenValid = nextIndexTokenAddress && allowedIndexTokens.has(nextIndexTokenAddress);
+  const isNextMarketTokenValid =
+    nextMarketTokenAdress && marketsMap[nextMarketTokenAdress]?.marketTokenAddress === nextMarketTokenAdress;
+
+  const fallbackPayToken = (fallbackPayToken?: string) => {
+    return produce(nextState, (draft) => {
+      draft.tokens.fromTokenAddress = fallbackPayToken ?? prevState.tokens.fromTokenAddress;
+    });
```

Comment:
I dont really undestand the benefit of of immer when its used to set a single field

Reply:
but it's nested single field, I think it's fine